server/commands/list_objects.go

```diff
@@ -0,0 +1,279 @@
+	err := q.validateInput(ctx, req.StoreId, req.Type, req.AuthorizationModelId, req.Relation)
+	if err != nil {
+		return err
+	}
+
+	resultsChan := make(chan string, q.ListObjectsMaxResults)
+	errChan := make(chan error)
+	resolvedChan := make(chan struct{})
+
+	timeoutCtx := ctx
+	if q.ListObjectsDeadline != 0 {
+		var cancel context.CancelFunc
+		timeoutCtx, cancel = context.WithTimeout(ctx, q.ListObjectsDeadline)
+		defer cancel()
+	}
+
+	go func() {
+		q.performChecks(timeoutCtx, &PerformChecksInput{
+			storeID:     req.StoreId,
+			authModelID: req.AuthorizationModelId,
+			objectType:  req.Type,
+			relation:    req.Relation,
+			user:        req.User,
+			ctxTuples:   req.ContextualTuples,
+		}, resultsChan, errChan, resolvedChan)
+	}()
+
+	for {
+		select {
+		case <-timeoutCtx.Done():
+			return nil
+		case objectID, ok := <-resultsChan:
+			if !ok {
+				return nil //channel was closed
+			}
+			if err := srv.Send(&openfgapb.StreamedListObjectsResponse{
+				ObjectId: objectID,
+			}); err != nil {
+				return serverErrors.NewInternalError("", err)
+			}
+		case genericError, ok := <-errChan:
+			if ok {
+				return serverErrors.NewInternalError("", genericError)
+			}
+		}
+	}
+}
+
+func (q *ListObjectsQuery) getUniqueObjects(ctx context.Context, storeID, targetObjectType string, ctxTuples *openfgapb.ContextualTupleKeys) (storage.ObjectIterator, error) {
+	iter, err := q.Datastore.ListObjectsByType(ctx, storage.ListObjectsFilter{
+		StoreID:    storeID,
+		ObjectType: targetObjectType,
+	})
+	if err != nil {
+		return nil, serverErrors.NewInternalError("", err)
+	}
+
+	return iter, nil
+}
+
+func (q *ListObjectsQuery) validateInput(ctx context.Context, storeID string, targetObjectType string, authModelID string, relation string) error {
+	definition, err := q.Datastore.ReadTypeDefinition(ctx, storeID, authModelID, targetObjectType)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return serverErrors.TypeNotFound(targetObjectType)
+		}
+		return err
+	}
+	_, ok := definition.Relations[relation]
+	if !ok {
+		return serverErrors.UnknownRelationWhenListingObjects(relation, targetObjectType)
+	}
+	return nil
+}
+
+type PerformChecksInput struct {
+	storeID     string
+	authModelID string
+	objectType  string
+	relation    string
+	user        string
+	ctxTuples   *openfgapb.ContextualTupleKeys
+}
+
+func (q *ListObjectsQuery) performChecks(timeoutCtx context.Context, input *PerformChecksInput, resultsChan chan<- string, errChan chan<- error, resolvedChan chan<- struct{}) {
+	g := new(errgroup.Group)
+	g.SetLimit(MaximumConcurrentChecks)
+	var objectsFound uint32
+
+	iter, err := q.getUniqueObjects(timeoutCtx, input.storeID, input.objectType, input.ctxTuples)
+	if err != nil {
+		errChan <- err
+		close(errChan)
+		return
+	}
+	defer iter.Stop()
+
+	// iterate over the contextual tuples
+	if input.ctxTuples != nil {
+		for _, t := range input.ctxTuples.TupleKeys {
```

Comment:
See https://github.com/openfga/openfga/pull/145/files#r943606933

We abstract this away in the `getUniqueObjects` method, especially if we're going to pass the contextual tuples to that function.